core/src/ModelArray.cpp

```diff
@@ -0,0 +1,332 @@
+{
+    ModelArray result(type, m_name + "-" + subtrahend.m_name);
+
+    for (size_t i = 0; i < m_sz.at(type); ++i) {
+        result[i] = (*this)[i] - subtrahend[i];
+    }
+
+    return result;
+}
+
+ModelArray ModelArray::operator*(const ModelArray& multiplier) const
+{
+    ModelArray result(type, m_name + "*" + multiplier.m_name);
+
+    for (size_t i = 0; i < m_sz.at(type); ++i) {
+        result[i] = (*this)[i] * multiplier[i];
+    }
+
+    return result;
+}
+
+ModelArray ModelArray::operator/(const ModelArray& divisor) const
+{
+    ModelArray result(type, m_name + "/" + divisor.m_name);
+
+    for (size_t i = 0; i < m_sz.at(type); ++i) {
+        result[i] = (*this)[i] / divisor[i];
+    }
+
+    return result;
+}
+
+ModelArray ModelArray::operator-() const
+{
+    ModelArray copy(type, std::string("-") + m_name);
+    copy.m_data = -m_data;
+    return copy;
+}
+
+ModelArray ModelArray::operator+(const double& x) const
+{
+    ModelArray copy(type, m_name + "+" + std::to_string(x));
+    copy.m_data = m_data + x;
+    return copy;
+}
+
+ModelArray ModelArray::operator-(const double& x) const
+{
+    ModelArray copy(type, m_name + "-" + std::to_string(x));
+    copy.m_data = m_data - x;
+    return copy;
+}
+
+ModelArray ModelArray::operator*(const double& x) const
+{
+    ModelArray copy(type, m_name + "*" + std::to_string(x));
+    copy.m_data = m_data * x;
+    return copy;
+}
+
+ModelArray ModelArray::operator/(const double& x) const
+{
+    ModelArray copy(type, m_name + "/" + std::to_string(x));
+    copy.m_data = m_data / x;
+    return copy;
+}
+
+ModelArray operator+(const double& x, const ModelArray& y) { return y + x; }
+
+ModelArray operator-(const double& x, const ModelArray& y) { return -(y - x); }
+
+ModelArray operator*(const double& x, const ModelArray& y) { return y * x; }
+
+ModelArray operator/(const double& x, const ModelArray& y)
+{
+    ModelArray xArray(y.getType(), std::to_string(x));
+    xArray.setData(x);
+    return xArray / y;
+}
+
+void ModelArray::setData(double value)
+{
+    resize();
+    m_data = value;
+}
+
+void ModelArray::setData(const double* pData)
+{
+    resize();
+    auto out = std::copy(pData, pData + m_sz.at(type), m_data.data());
+}
+
+void ModelArray::setData(const DataType& from) { setData(from.data()); }
+
+void ModelArray::setData(const ModelArray& from) { setData(from.m_data.data()); }
+
+void ModelArray::setDimensions(Type type, const Dimensions& newDims)
+{
+    size_t newSize = 1;
+    for (size_t dimLen : newDims) {
+        newSize *= dimLen;
+    }
+    m_dims.at(type).clear();
+    std::copy(newDims.begin(), newDims.end(), std::back_inserter(m_dims.at(type)));
+    m_sz.at(type) = newSize;
+}
+
+// Simple 1d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i) { return i; }
+
+// Special case for 2d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i, size_t j)
+{
+    return i * dims[1] + j;
+}
+
```

Comment:
If this function is supposed to be implementing linear indexing in multi-dimensional arrays, I think it produces wrong results for more than 3 dimensions. Also, I suggest you variadic templates to implement this. For example:

```
// Base case
template<typename T>
T index(const T *dims, T ndims, T v) {
  return v;
}

template <typename T, typename... Args>
T index(const T *dims, T ndims, T first, Args... args)
{
    // FIXME: check number of dimensions match number of indices
    T dim_prod = 1;
    for (int i = 0; i < ndims - 1; i++)
        dim_prod *= dims[ndims - i - 1];
    return first*dim_prod + index(dims, ndims - 1, args...);
}

```
There is a better way to implement this without recursion.